pkg/tbtc/node.go

```diff
@@ -1075,6 +1089,172 @@ func processCoordinationResult(node *node, result *coordinationResult) {
 	}
 }
 
+// archiveClosedWallets archives closed or terminated wallets.
+func (n *node) archiveClosedWallets() error {
+	getClosedWallets := func(walletPublicKeyHashes [][20]byte) (
+		closedWallets [][20]byte,
+	) {
+		for _, walletPublicKeyHash := range walletPublicKeyHashes {
+			walletChainData, err := n.chain.GetWallet(walletPublicKeyHash)
+			if err != nil {
+				// Continue if there was an error getting wallet data. Try to
+				// get as many closed wallets as possible.
+				logger.Errorf(
+					"could not get wallet data for wallet [0x%x]: [%v]",
+					walletPublicKeyHash,
+					err,
+				)
+				continue
+			}
+
+			if walletChainData.State == StateClosed ||
+				walletChainData.State == StateTerminated {
+				closedWallets = append(closedWallets, walletPublicKeyHash)
+			}
+		}
+
+		return
+	}
+
+	// Get all the wallets controlled by the node.
+	walletPublicKeys := n.walletRegistry.getWalletsPublicKeys()
+
+	walletPublicKeyHashes := [][20]byte{}
+	for _, walletPublicKey := range walletPublicKeys {
+		walletPublicKeyHashes = append(
+			walletPublicKeyHashes,
+			bitcoin.PublicKeyHash(walletPublicKey),
+		)
+	}
+
+	// Find the wallets that are closed.
+	initialClosedWallets := getClosedWallets(walletPublicKeyHashes)
+	if len(initialClosedWallets) == 0 {
+		logger.Infof("there are no wallets to be archived")
+		return nil
+	}
+
+	// Wait a significant number of blocks to make sure the transactions that
+	// caused the wallets to be closed have not been reverted for some reason,
+	// e.g. due to a chain reorganization.
+	ctx := context.Background()
+
+	blockCounter, err := n.chain.BlockCounter()
+	if err != nil {
+		return fmt.Errorf("error getting block counter [%v]", err)
+	}
+
+	currentBlock, err := blockCounter.CurrentBlock()
+	if err != nil {
+		return fmt.Errorf("error getting current block [%v]", err)
+	}
+
+	confirmationBlock := currentBlock + walletClosureConfirmationBlocks
```

Comment:
I understand the purpose but:
- This will block client startup for 32 blocks (~6 min) which is bad
- I'm not sure if we actually need this check here. There is a very low probability that a client would restart roughly at the same time a wallet closure transaction gets reorganized and rejected by the chain. A wallet closure is quite a rare event. A reorg that would evict it from the chain is even rarer as this would require another transaction entering just before the closure transaction and changing the state in a way that would revert the closure. Also, even if such a scenario occurs, it would affect only the restarting client so the impact is limited. Last but not least, this is easily recoverable as we can always release a patch to unarchive the key material.

I propose to remove the confirmation check from here and simplify this code. Of course, the confirmation of the trigger based on `WalletClosed` events should stay. This is a different case with a higher probability and potential impact (all nodes are listening for those events at the same time).

Reply:
Makes sense.
Done in 442905743cf198b3ee3a602ad180951898432acf.